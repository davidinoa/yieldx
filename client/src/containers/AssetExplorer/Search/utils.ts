interface InGridFilter {
  filter: string | number;
  filterTo: number;
  filterType: string;
  type: string;
  dateFrom: string;
  dateTo: string;
  values: string[];
}

export interface AgGridFilters {
  [key: string]: any;
}

const getFilter = (filter: InGridFilter) => {
  /**
   *  The filter object from Ag Grid API. In the case of filtering `name` column.
   *  { filter: "Asset Name", filterType: "text", type: "contains"}
   */
  let searchFilter = {};
  const {
    filter: filterContent,
    filterType,
    type,
    filterTo,
    values,
    dateFrom,
    dateTo
  } = filter;
  if (filterType === "text") {
    switch (type) {
      case "contains":
        searchFilter = { containsText: filterContent };
        break;
      case "notContains":
        searchFilter = { doesNotContainText: filterContent };
        break;
      case "equals":
        searchFilter = { equalsText: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualText: filterContent };
        break;
      case "startsWith":
        searchFilter = { startsWithText: filterContent };
        break;
      case "endsWith":
        searchFilter = { endsWith: filterContent };
        break;
      default:
        searchFilter = { containsText: filterContent };
        break;
    }
  } else if (filterType === "number") {
    switch (type) {
      case "equals":
        searchFilter = { equalsNumber: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualNumber: filterContent };
        break;
      case "lessThan":
        searchFilter = { isLessThanNumber: filterContent };
        break;
      case "lessThanOrEqual":
        searchFilter = { isLessThanOrEqualToNumber: filterContent };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanNumber: filterContent };
        break;
      case "greaterThanOrEqual":
        searchFilter = { isGreaterThanOrEqualToNumber: filterContent };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToNumber: filterContent,
          isLessThanOrEqualToNumber: filterTo
        };
        break;
      default:
        searchFilter = { equalsNumber: filterContent };
        break;
    }
  } else if (filterType === "date") {
    const convertedDateFrom = new Date(dateFrom);
    const convertedDateTo = new Date(dateTo);
    switch (type) {
      case "equals":
        searchFilter = { equalsDate: convertedDateFrom };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanDate: convertedDateFrom };
        break;
      case "lessThan":
        searchFilter = { isLessThanDate: convertedDateFrom };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualDate: convertedDateFrom };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToDate: convertedDateFrom,
          isLessThanOrEqualToDate: convertedDateTo
        };
        break;
      default:
        searchFilter = { equalsDate: convertedDateFrom };
        break;
    }
  } else if (filterType === "set" && values) {
    // We may need to add `set` type Ag grid column filter params like cusip or isin
    // in the search request in the future.
    // searchFilter = { inText: values };
  }
  return searchFilter;
};

export const convertInGridFilters = (filters: AgGridFilters) => {
  const convertedFilters: { [index: string]: any } = {};
  Object.keys(filters).forEach(key => {
    const inGridFilter = filters[key];
    if (key === "minimumQuantity") key = "bond.minimumDenomination";
    convertedFilters[key] = getFilter(inGridFilter);
  });

  return convertedFilters;
};
