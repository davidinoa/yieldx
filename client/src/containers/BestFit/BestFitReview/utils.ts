import { getRatingValue } from "../../../util/ratings";
import YieldXPortfolio, {
  YieldXPosition
} from "../../../models/YieldXPortfolio";

export function getBondRatingsData(
  portfolio: YieldXPortfolio
): { name: string; value: number }[] {
  if (portfolio.getBondSleeve) {
    const ratingMarketValues: { [key: string]: number } = {};
    const bondPortfolio = portfolio.getBondSleeve();
    bondPortfolio.positions.forEach((position: YieldXPosition) => {
      const { asset, weight } = position;
      const rating = asset.rating ?? "NR";
      ratingMarketValues[rating] = (ratingMarketValues[rating] || 0) + weight;
    });
    return Object.entries(ratingMarketValues)
      .map(([rating, weight]) => {
        return {
          name: rating,
          value: weight
        };
      })
      .sort((a, b) => {
        return getRatingValue(a.name) - getRatingValue(b.name);
      });
  }
  return [];
}
