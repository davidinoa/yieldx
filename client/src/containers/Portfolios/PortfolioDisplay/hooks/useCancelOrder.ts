import {
  OrderStatus,
  usePatchPortfolioOrderCancelMutation
} from "providers/graphql/hooks";

function useCancelOrder() {
  const [cancelOrder] = usePatchPortfolioOrderCancelMutation();
  return (id: string, portfolioId: string) =>
    cancelOrder({
      variables: {
        id,
        portfolioId,
        cancelOrderRequestInput: {}
      },
      update(cache) {
        cache.modify({
          id: `Order2:${id}`,
          fields: {
            status() {
              return OrderStatus.Canceled;
            }
          }
        });
      }
    });
}

export default useCancelOrder;
