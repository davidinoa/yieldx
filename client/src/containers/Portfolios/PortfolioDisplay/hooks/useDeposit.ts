import {
  usePostInvestorPortfolioInternalPortfolioFundTransferMutation,
  Position2
} from "providers/graphql/hooks";

import YieldXPortfolio from "models/YieldXPortfolio";

interface Params {
  investorId: string;
  amount: number;
  cashPortfolio: {
    cashPos: Position2;
    cashAvailable: number;
    portfolioId: string;
  };
  portfolio: YieldXPortfolio;
}

function useDeposit() {
  const [
    makeTransfer
  ] = usePostInvestorPortfolioInternalPortfolioFundTransferMutation();

  return ({ investorId, amount, portfolio, cashPortfolio }: Params) => {
    const { cashAvailable } = cashPortfolio;
    const cashPosition = portfolio.positions.find(
      position => position.id === "CASH-USD" || position.id === "USD"
    );
    const cashAvailableInPortfolio =
      (cashPosition?.unsettled ?? 0) > 0
        ? cashPosition?.settled ?? 0
        : cashPosition?.quantity ?? 0;

    const addTransferPortfolio = cashAvailableInPortfolio + amount;
    const subtractTransfer = cashAvailable - amount;

    makeTransfer({
      variables: {
        investorId,
        portfolioId: portfolio.getId(),
        createFundTransferRequestInput: {
          quantity: amount,
          currency: "USD"
        }
      },
      update(cache) {
        cache.modify({
          id: `Position2:USD:${portfolio.getId()}`,
          fields: {
            settled() {
              return addTransferPortfolio;
            },
            quantity() {
              return addTransferPortfolio;
            }
          }
        });

        cache.modify({
          id: `Position2:USD:${cashPortfolio.portfolioId}`,
          fields: {
            settled() {
              return subtractTransfer;
            },
            quantity() {
              return subtractTransfer;
            }
          }
        });
      }
    });
  };
}

export default useDeposit;
