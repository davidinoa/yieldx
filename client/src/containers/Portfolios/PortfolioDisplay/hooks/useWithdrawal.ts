import {
  usePostInvestorPortfolioInternalPortfolioFundTransferMutation,
  Position2
} from "providers/graphql/hooks";

import YieldXPortfolio from "models/YieldXPortfolio";

interface Params {
  investorId: string;
  amount: number;
  cashPortfolio: {
    cashPos: Position2;
    cashAvailable: number;
    portfolioId: string;
  };
  portfolio: YieldXPortfolio;
}

function useWithdrawal() {
  const [
    makeTransfer
  ] = usePostInvestorPortfolioInternalPortfolioFundTransferMutation();

  return ({ investorId, amount, portfolio, cashPortfolio }: Params) => {
    const { cashAvailable } = cashPortfolio;
    const cashPosition = portfolio.positions.find(
      position => position.id === "CASH-USD" || position.id === "USD"
    );
    const cashAvailableInPortfolio =
      (cashPosition?.unsettled ?? 0) > 0
        ? cashPosition?.settled ?? 0
        : cashPosition?.quantity ?? 0;

    const addTransfer = cashAvailable + amount;
    const subtractTransferPortfolio = cashAvailableInPortfolio - amount;
    makeTransfer({
      variables: {
        investorId,
        portfolioId: portfolio.getId(),
        createFundTransferRequestInput: {
          quantity: amount * -1,
          currency: "USD"
        }
      },
      update(cache) {
        cache.modify({
          id: `Position2:USD:${portfolio.getId()}`,
          fields: {
            settled() {
              return subtractTransferPortfolio;
            },
            quantity() {
              return subtractTransferPortfolio;
            }
          }
        });

        cache.modify({
          id: `Position2:USD:${cashPortfolio.portfolioId}`,
          fields: {
            settled() {
              return addTransfer;
            },
            quantity() {
              return addTransfer;
            }
          }
        });
      }
    });
  };
}

export default useWithdrawal;
