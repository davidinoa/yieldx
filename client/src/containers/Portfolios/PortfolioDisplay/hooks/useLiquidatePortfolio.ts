import {
  Direction2,
  Order2,
  OrderManagementGetOrderDocument,
  OrderStatus,
  PortfolioStatus,
  usePatchPortfolioMutation,
  usePostPortfolioOrdersMutation
} from "providers/graphql/hooks";

import YieldXPortfolio from "models/YieldXPortfolio";
import useCancelOrder from "containers/Portfolios/PortfolioDisplay/hooks/useCancelOrder";

function useLiquidatePortfolio() {
  const [updatePortfolio] = usePatchPortfolioMutation();
  const [postOrder] = usePostPortfolioOrdersMutation();
  const cancelOrder = useCancelOrder();

  return async (portfolio: YieldXPortfolio, orders: Array<Order2>) => {
    orders.forEach(order => {
      const { id: orderId, status } = order ?? {};
      if (!orderId) throw Error("No order ID");
      if (status === OrderStatus.Executing)
        throw Error("Cannot liquidate portfolio with executing orders");
      if (status === OrderStatus.Open) cancelOrder(orderId, portfolio.getId());
    });
    const referencePricesList =
      portfolio?.positions?.map(it => it.asset.lastPrice) ?? [];
    const referencePricesMap = referencePricesList.reduce<{
      [assetId: string]: number | null;
    }>((acc, cv, i) => {
      acc[portfolio?.positions[i]?.assetId ?? ""] = !cv ? null : cv;
      return acc;
    }, {});
    await Promise.all(
      portfolio.positions.map(position =>
        postOrder({
          variables: {
            portfolioId: portfolio.getId(),
            createOrderRequestInput: {
              assetId: position.assetId,
              quantity: position.quantity,
              goodTillCancel: true,
              direction: Direction2.Sell,
              priceLimit: (position.asset?.price ?? 1) * 0.98,
              spotPrice:
                referencePricesMap[position.assetId] ??
                position.asset.price ??
                undefined
            }
          },
          update(cache, { data }) {
            const newOrder = data?.postPortfolioOrders?.order;
            if (newOrder)
              cache.writeQuery({
                data: newOrder,
                query: OrderManagementGetOrderDocument,
                variables: {
                  id: newOrder.id,
                  portfolioId: newOrder.portfolioId
                }
              });
          }
        })
      )
    );

    await updatePortfolio({
      variables: {
        id: portfolio.getId(),
        updatePortfolioRequestInput: {
          status: PortfolioStatus.Terminated
        }
      },
      update(cache) {
        cache.modify({
          id: `Portfolio3:${portfolio.getId()}`,
          fields: {
            status() {
              return PortfolioStatus.Terminated;
            }
          }
        });
      }
    });
  };
}

export default useLiquidatePortfolio;
