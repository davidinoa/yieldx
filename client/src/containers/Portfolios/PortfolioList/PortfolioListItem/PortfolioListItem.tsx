import React from "react";
import { AppType2, Portfolio3, PortfolioType } from "providers/graphql/hooks";
import {
  CTAColumn,
  DetailsButton,
  ListRow,
  Status,
  StatusColumn
} from "components/ListPages/ListPages.styles";
import dayjs from "dayjs";
import { Spinner } from "elements/Spinner/Spinner";
import { ReactComponent as RedAlert } from "assets/svg/red-alert.svg";
import useGetPortfolioInvestor from "providers/graphql/hooks/useGetPortfolioInvestor";
import { List } from "components/ListPages/List";

interface Props {
  portfolio: Portfolio3;
  portfolioType: PortfolioType;
}

function PortfolioListItem({ portfolio, portfolioType }: Props) {
  const { id, name, createdAt } = portfolio;
  const isExternal = portfolioType === PortfolioType.External;
  const status = (isExternal ? "tracking" : portfolio.status).toLowerCase();
  const { investor, error, loading } = useGetPortfolioInvestor(
    id,
    portfolioType !== PortfolioType.Live
  );
  return (
    <ListRow key={id} status={status}>
      <StatusColumn>
        <Status className={status} />
      </StatusColumn>
      <List.Col
        className="col-14 col-xxl-16"
        label={
          investor
            ? `${investor?.investorInfo.givenName} ${investor?.investorInfo.familyName}`
            : undefined
        }
        value={name}
      >
        {(error || null) && <RedAlert />}
        {(loading || null) && <Spinner />}
      </List.Col>
      <List.Col
        className="col-2 centered"
        label="type"
        value={portfolio.app === AppType2.Inpaas ? "InPaaS" : "Custom"}
      />
      <List.Col
        className="col-4 col-xxl-3 centered"
        value={dayjs(createdAt).format("MMM D, YYYY")}
        label="Created at"
      />
      <CTAColumn>
        <DetailsButton
          data-testid="PortfolioList-ShowPortfolioLink"
          to={`/portfolios/${isExternal ? "external" : "live"}/${portfolio.id}`}
        >
          Details
        </DetailsButton>
      </CTAColumn>
    </ListRow>
  );
}

export default PortfolioListItem;
