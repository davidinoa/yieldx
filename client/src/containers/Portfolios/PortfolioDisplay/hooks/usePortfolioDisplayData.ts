import { useHistory, useLocation, useParams } from "react-router-dom";
import {
  Order2,
  PortfolioType,
  Position2,
  useOrderManagementGetOrdersQuery
} from "providers/graphql/hooks";
import useGetYieldxPortfolio from "providers/graphql/hooks/useGetYieldxPortfolio";
import useGetPortfolioInvestor from "providers/graphql/hooks/useGetPortfolioInvestor";
import useGetInvestorCashPortfolio from "providers/graphql/hooks/useGetInvestorCashPortfolio";
import useCancelOrder from "containers/Portfolios/PortfolioDisplay/hooks/useCancelOrder";
import useDeleteExternalPortfolio from "providers/graphql/hooks/useDeleteExternalPortfolio";
import useLiquidatePortfolio from "containers/Portfolios/PortfolioDisplay/hooks/useLiquidatePortfolio";
import useDeposit from "containers/Portfolios/PortfolioDisplay/hooks/useDeposit";
import useWithdrawal from "containers/Portfolios/PortfolioDisplay/hooks/useWithdrawal";
import {
  AlertLevel,
  AlertMessage,
  useAlertContext
} from "providers/alerts/AlertProvider";

type PortfolioDisplayParams = {
  portfolioId: string;
  type: string;
};

export default function usePortfolioDisplayData() {
  const { type, portfolioId } = useParams<PortfolioDisplayParams>();
  const portfolioType =
    type === "external" ? PortfolioType.External : PortfolioType.Live;

  const history = useHistory();
  const { pathname, hash } = useLocation();

  const {
    portfolio,
    loading: portfolioLoading,
    error: portfolioError
  } = useGetYieldxPortfolio({ portfolioId, portfolioType });
  if (portfolioError) console.error(portfolioError);
  const {
    data: ordersData,
    loading: ordersLoading,
    error: ordersError
  } = useOrderManagementGetOrdersQuery({
    variables: {
      portfolioId
    },
    skip: portfolioType === "EXTERNAL" || !portfolio
  });
  const orders = ordersData?.orderManagementGetOrders?.orders;
  const cancelOrder = useCancelOrder();
  const deleteExternal = useDeleteExternalPortfolio();
  const {
    investor,
    error: investorError,
    loading: investorLoading
  } = useGetPortfolioInvestor(
    portfolioId,
    !portfolio || portfolioType !== PortfolioType.Live
  );

  const {
    data: { cashPortfolio },
    error: cashPortfolioError,
    loading: cashPortfolioLoading
  } = useGetInvestorCashPortfolio(investor?.id);

  const liquidatePortfolio = useLiquidatePortfolio();
  const makeDeposit = useDeposit();
  const makeWithdrawal = useWithdrawal();
  const { triggerAlert } = useAlertContext();
  return {
    portfolio,
    portfolioLoading,
    portfolioError,
    orders,
    ordersLoading,
    ordersError,
    investorError,
    investorLoading,
    cashPortfolioError,
    cashPortfolioLoading,
    history,
    pathname,
    hash,
    investorCashAvailable: cashPortfolio?.cashAvailable,
    onDelete: async () => {
      try {
        await deleteExternal(portfolioId).then(() => {
          triggerAlert(
            "You portfolio has been closed successfully",
            AlertLevel.success,
            { timeout: 2500 }
          );
        });
      } catch (err) {
        console.error(err);
        triggerAlert(AlertMessage.Unexpected, AlertLevel.danger);
      }
      history.push("/portfolios#alert=PortfolioDeleted");
    },
    onCancelOrder: async (orderId: string) => {
      try {
        await cancelOrder(orderId, portfolioId);
        history.push(`${pathname}#orders`);
      } catch (e) {
        console.error(e);
      }
    },
    onLiquidate: async () => {
      const investorId = investor?.id;
      if (!investorId) throw Error("Investor ID is missing");
      if (!portfolio)
        throw Error("Portfolio is undefined. Failed to liquidate");
      if (!orders) throw Error("Orders undefined, failed to liquidate");
      await liquidatePortfolio(portfolio, orders as Array<Order2>);
      history.push(`/investors/${investorId}#portfolios`, {
        shouldRefresh: true
      });
    },

    onDeposit: async (amount: number) => {
      const investorId = investor?.id;
      if (!investorId) throw Error("Investor ID is missing");
      if (!portfolio) throw Error("Portfolio missing");
      if (!cashPortfolio) throw Error("Cash investor/portfolio is missing");
      makeDeposit({
        investorId,
        portfolio,
        amount,
        cashPortfolio: cashPortfolio as {
          cashPos: Position2;
          cashAvailable: number;
          portfolioId: string;
        }
      });
      history.push(pathname);
    },

    onWithdraw: async (amount: number) => {
      const investorId = investor?.id;
      if (!investorId) throw Error("Investor ID is missing");
      if (!portfolio) throw Error("Portfolio missing");
      if (!cashPortfolio) throw Error("Cash investor/portfolio is missing");
      makeWithdrawal({
        investorId,
        portfolio,
        amount,
        cashPortfolio: cashPortfolio as {
          cashPos: Position2;
          cashAvailable: number;
          portfolioId: string;
        }
      });
      history.push(pathname);
    }
  };
}
