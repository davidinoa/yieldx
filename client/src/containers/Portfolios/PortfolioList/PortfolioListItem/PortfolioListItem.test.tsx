import React from "react";
import { render } from "test/utils";
import PortfolioListItem from "containers/Portfolios/PortfolioList/PortfolioListItem/PortfolioListItem";
import mockPortfolio from "test/mock-data/mockPortfolio";
import { PortfolioType } from "providers/graphql/hooks";

describe("PortfolioListItem", () => {
  it("It should render external portfolio properly", () => {
    const { asFragment } = render(
      <PortfolioListItem
        portfolio={mockPortfolio({ createdAt: "11/05/2020" })}
        portfolioType={PortfolioType.External}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
