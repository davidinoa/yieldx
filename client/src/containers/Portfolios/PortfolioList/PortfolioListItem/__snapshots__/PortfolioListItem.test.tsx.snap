// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PortfolioListItem It should render external portfolio properly 1`] = `
<DocumentFragment>
  .c3 {
  font-weight: bold;
  color: var(--text);
  font-size: 0.725rem;
  margin: 0;
  line-height: normal;
}

.c4 {
  margin: 0;
  font-weight: bold;
  font-size: 0.6rem;
  -webkit-letter-spacing: 0.2em;
  -moz-letter-spacing: 0.2em;
  -ms-letter-spacing: 0.2em;
  letter-spacing: 0.2em;
  text-transform: uppercase;
  color: var(--subtext-3);
}

.c0 {
  height: 70px;
  list-style: none;
  border: 1px solid var(--light);
  background: var(--dark);
  border-radius: 5px;
  position: relative;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  margin-bottom: 0.45rem;
}

.c0 .no-borders::after {
  border-right: none !important;
  border-left: none !important;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 5px;
  width: 5px;
  border-radius: 50%;
  margin: 0 auto;
}

.c2.active,
.c2.ready {
  background: var(--active);
}

.c2.deactivated,
.c2.inactive,
.c2.blocked,
.c2.closed,
.c2.terminated {
  background: var(--subtext-3);
}

.c2.tracking {
  background: var(--tracking);
}

.c2.pending {
  background: var(--caution);
}

.c2.error {
  background: var(--issue);
}

.c5 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: var(--page);
  font-weight: bold;
  border-radius: 0.25rem;
  color: white;
  -webkit-text-decoration: none;
  text-decoration: none;
  width: 70px;
  height: 26px;
  font-size: 0.8rem;
}

.c5:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: white;
}

@media (min-width:1600px) {
  .c3 {
    font-size: 0.875rem;
  }
}

@media (min-width:1600px) {
  .c4 {
    font-size: 0.6875rem;
  }
}

@media (min-width:1600px) {
  .c0 {
    padding-top: 0.75rem;
    padding-bottom: 0.75rem;
    margin-bottom: 0.575rem;
  }
}

@media (min-width:1600px) {
  .c2 {
    height: 6px;
    width: 6px;
  }
}

@media (min-width:1600px) {
  .c5 {
    width: 90px;
    height: 30px;
    font-size: 0.975rem;
  }
}

<div
    class="c0 row"
    status="tracking"
  >
    <div
      class="c1 col-1 no-borders"
    >
      <i
        class="c2 tracking"
      />
    </div>
    <div
      class="d-flex flex-column justify-content-center col-14 col-xxl-16 col"
    >
      <span
        class="c3"
      >
        MOCK PORTFOLIO
      </span>
    </div>
    <div
      class="d-flex flex-column justify-content-center col-2 centered col"
    >
      <span
        class="c3"
      >
        InPaaS
      </span>
      <span
        class="c4"
      >
        type
      </span>
    </div>
    <div
      class="d-flex flex-column justify-content-center col-4 col-xxl-3 centered col"
    >
      <span
        class="c3"
      >
        Nov 5, 2020
      </span>
      <span
        class="c4"
      >
        Created at
      </span>
    </div>
    <div
      class="d-flex flex-column col-3 col-xxl-2 centered justify-content-center"
    >
      <a
        class="c5"
        data-testid="PortfolioList-ShowPortfolioLink"
        href="/portfolios/external/mock-portfolio-id"
      >
        Details
      </a>
    </div>
  </div>
</DocumentFragment>
`;
